### PR TITLE
Document the can-block-on relationship for servo.

### LINKED ISSUE #14704: Document deadlock-freedom of the constellation, since IPC buffer send is blocking
If a script thread busy-loops, but the constellation keeps sending it messages, eventually it will exhaust the IPC buffer. This will then block the constellation, oh dear.

IPC conversation with @antrik and @pcwalton http://logs.glob.uno/?c=mozilla%23servo&s=23+Dec+2016&e=23+Dec+2016#c583154

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR adds some documentation comments describing the can-block-on relation that is used to ensure deadlock-freedom.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14704 
- [X] These changes do not require tests because it's documentation

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18810)

### CODE REVIEW

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -64,6 +64,34 @@
...
+//! * Layout can block on font cache
+//! * Layout can block on image cache
+//! * Constellation can block on compositor
+//! * Constellation can block on embedder
+//! * Constellation can block on layout
+//! * Script can block on anything (other than script)
+//! * Blocking is transitive (if T1 can block on T2 and T2 can block on T3 then T1 can block on T3)
+//! * Nothing can block on itself!
+//!
+//! There is a wrinkle around the use of IPC channels, since they do not support
```
Comment (nox) [inline components/constellation/constellation.rs]: There is a what? Please reword.

Comment (asajeffrey) [reply components/constellation/constellation.rs]: Done.

File: components/constellation/constellation.rs [hunk 2]
```diff
@@ -64,6 +64,34 @@
...
+//! * Layout can block on image cache
+//! * Constellation can block on compositor
+//! * Constellation can block on embedder
+//! * Constellation can block on layout
+//! * Script can block on anything (other than script)
+//! * Blocking is transitive (if T1 can block on T2 and T2 can block on T3 then T1 can block on T3)
+//! * Nothing can block on itself!
+//!
+//! There is a complexity intoduced by IPC channels, since they do not support
+//! non-blocking send. This means that as well as `receiver.recv()` blocking, so
```
Comment (cbrewster) [inline components/constellation/constellation.rs]: nit: This means that as well as `receiver.recv()` blocking, `sender.send(data)` can also block when the IPC buffer is full.

**Review by cbrewster — APPROVED** [review]
LGTM.

Glad to have this documented.

File: components/constellation/constellation.rs (hunk 2 above)
Comment (asajeffrey) [reply components/constellation/constellation.rs]: Fixed.

**Pull request merged**